File: 2020/day-6/aoc-6.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Sun Dec  6 12:20:04 2020

@author: mattiaruefenacht
"""

import os
from typing import List
from itertools import chain

os.getcwd()

os.chdir(r'/Users/mattiaruefenacht/Desktop/AOC2020/day-6')

def grouping() -> List[List[str]]:
    with open("input.txt") as doc:
        return [group.split("\n") for group in doc.read().split("\n\n")]

def count_unique(groups): 
    return sum(len(set(chain(*group))) for group in groups)

def count_common(groups):
    return sum(len(set.intersection(*(set(person) for person in group)))
               for group in groups)

def main():
    input_groups = grouping()
    print('Part 1 solution:', count_unique(input_groups))
    print('Part 2 solution:', count_common(input_groups))

if __name__ == "__main__":
    main()

File: 2020/day-6/aoc-6.go
package main

import (
	"fmt"
	"io/ioutil"
	"strings"
)

func main() {
	data, err := ioutil.ReadFile("input.txt")
	if err != nil {
		panic(err)
	}
	contents := string(data)
	split := strings.Split(contents, "\n")
	split = split[:len(split)-1]
	groupNumber := 0
	groups := []map[rune]int{
		make(map[rune]int),
	}
	groupSize := []int{0}
	for _, s := range split {
		if s == "" {
			groups = append(groups, make(map[rune]int))
			groupSize = append(groupSize, 0)
			groupNumber++
			continue
		}
		for _, r := range s {
			groups[groupNumber][r]++
		}
		groupSize[groupNumber]++
	}
	sum := 0
	yes := 0
	for n, g := range groups {
		sum += len(g)
		for _, count := range g {
			if count == groupSize[n] {
				yes++
			}
		}
	}
	fmt.Println("Part 1 solution:", sum)
	fmt.Println("Part 2 solution:", yes)
}
